### PR TITLE
Typeahead search control (static & dynamic) in Adaptive Cards - C# sample

### PR DESCRIPTION
**Pull request by Prithvi-MSFT**

### CODE REVIEW

File: samples/type_ahead_search_adaptive_cards/csharp/README.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
+---
+page_type: sample
+description: Demonstrating feature of type ahead search (static and dynamic) control in Adaptive cards.
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/README.md:3]: Demonstrating the feature of type ahead search (static and dynamic) control in Adaptive Cards.

File: samples/type_ahead_search_adaptive_cards/csharp/README.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+- office-365
+languages:
+- nodejs
+extensions:
+contentType: samples
+createdDate: "11-11-2021 23:30:17"
+---
+# Type ahead search control in Adaptive cards
+
+This sample shows feature of type ahead search (static and dynamic) control in Adaptive cards.
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/README.md:16]: This sample shows the feature of type ahead search (static and dynamic) control in Adaptive Cards.

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  "id": "{{Microsoft-App-Id}}",
+  "packageName": "com.microsoft.teams.typeaheadsearchcard",
+  "developer": {
+    "name": "Microsoft",
+    "websiteUrl": "https://www.microsoft.com",
+    "privacyUrl": "https://www.microsoft.com/privacy",
+    "termsOfUseUrl": "https://www.microsoft.com/termsofuse"
+  },
+  "name": {
+    "short": "Search adaptive card",
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json:14]: Type ahead search on Adaptive Card

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    "full": "Adaptive card with search control"
+  },
+  "description": {
+    "short": "Search control adaptive card",
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json:18]: Type ahead search on Adaptive Card

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    "full": "This sample shows feature type ahead search (static and dynamic) control in adaptive cards."
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json:19]: This sample shows the feature of type ahead search (static and dynamic) control on Adaptive Cards.

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      ],
+      "commandLists": [
+        {
+          "scopes": [
+            "personal"
+          ],
+          "commands": [
+            {
+              "title": "staticsearch",
+              "description": "To get static search card"
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json:41]: To get static type ahead search Adaptive Card

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            },
+            {
+              "title": "dynamicsearch",
+              "description": "To get dynamic search card"
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json:45]: To get dynamic type ahead search Adaptive Card

File: samples/type_ahead_search_adaptive_cards/csharp/README.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
+---
+page_type: sample
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/README.md:2]: Please link this readme file with root folder readme file.

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        /// <returns>A task that represents the work queued to execute.</returns>
+        protected override async Task<InvokeResponse> OnInvokeActivityAsync(ITurnContext<IInvokeActivity> turnContext, CancellationToken cancellationToken)
+        {
+            if (turnContext.Activity.Name == "application/search")
+            {
+                var searchData = JsonConvert.DeserializeObject<DynamicSearchCard>(turnContext.Activity.Value.ToString());
+                var packageResult = JObject.Parse(await (new HttpClient()).GetStringAsync($"https://azuresearch-usnc.nuget.org/query?q=id:{searchData.queryText}&prerelease=true"));
+                var packages = packageResult["data"].Select(item => (item["id"].ToString(), item["description"].ToString()));
+                var itemList = packages.Select(item => { var obj = new { Title = item.Item1, Value = item.Item2 }; return obj; }).ToList();
+                InvokeResponse adaptiveCardResponse;
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs:93]: Unused variable, please remove it.

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        protected override async Task<InvokeResponse> OnInvokeActivityAsync(ITurnContext<IInvokeActivity> turnContext, CancellationToken cancellationToken)
+        {
+            if (turnContext.Activity.Name == "application/search")
+            {
+                var searchData = JsonConvert.DeserializeObject<DynamicSearchCard>(turnContext.Activity.Value.ToString());
+                var packageResult = JObject.Parse(await (new HttpClient()).GetStringAsync($"https://azuresearch-usnc.nuget.org/query?q=id:{searchData.queryText}&prerelease=true"));
+                var packages = packageResult["data"].Select(item => (item["id"].ToString(), item["description"].ToString()));
+                var itemList = packages.Select(item => { var obj = new { Title = item.Item1, Value = item.Item2 }; return obj; }).ToList();
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs:92]: What is itemList ?
Rename it properly

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Cards/DynamicSearchCard.json [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+  ],
+  "actions": [
+    {
+      "type": "Action.Submit",
+      "id": "submitdynamic",
+      "title": "Submit"
+    }
+  ],
+  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+  "version": "1.2"
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Cards/DynamicSearchCard.json:70]: Put these 2 lines at top.

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Cards/StaticSearchCard.json [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+  ],
+  "actions": [
+    {
+      "type": "Action.Submit",
+      "id": "submit",
+      "title": "Submit"
+    }
+  ],
+  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+  "version": "1.2"
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Cards/StaticSearchCard.json:94]: Put these 2 lines at top

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Models/RequestDetails.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+        public string UserName { get; set; }
+
+        public string ManagerName { get; set; }
+
+        public string UserCardId { get; set; }
+
+        public string ManagerCardId { get; set; }
+        public string Status { get; set; }
+    }
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Models/RequestDetails.cs:25]: Why do we need this model??

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Startup.cs [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+            services.AddControllers().AddNewtonsoftJson();
+
+            services.AddHttpClient().AddControllers().AddNewtonsoftJson();
+            services.AddRazorPages();
+
+            // Create the Bot Framework Adapter with error handling enabled.
+            services.AddSingleton<IBotFrameworkHttpAdapter, AdapterWithErrorHandler>();
+
+            // Create the storage we'll be using for User and Conversation state. (Memory is great for testing purposes.)
+            services.AddSingleton<IStorage, MemoryStorage>();
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Startup.cs:39]: Where are we using this storage?

File: samples/type_ahead_search_adaptive_cards/csharp/README.md [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+- office-teams
+- office
+- office-365
+languages:
+- nodejs
+extensions:
+contentType: samples
+createdDate: "11-11-2021 23:30:17"
+---
+# Type ahead search control in Adaptive cards
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/README.md:14]: Type ahead search control on Adaptive Cards

File: samples/type_ahead_search_adaptive_cards/csharp/README.md [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+This sample shows feature of type ahead search (static and dynamic) control in Adaptive cards.
+
+`Static search`
+
+![static search card](Images/staticSearchCard.png)
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/README.md:20]: Images are not displaying properly on UI

File: samples/type_ahead_search_adaptive_cards/csharp/README.md [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+`Dynamic search`
+
+![dynamic search card](Images/dynamicSearchCard.png)
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/README.md:24]: Images are not displaying properly on UI

File: samples/type_ahead_search_adaptive_cards/csharp/README.md [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+  ```bash
+  ngrok http -host-header=localhost 3978
+  ```
+- [Teams](https://teams.microsoft.com) Microsoft Teams is installed and you have an account
+
+## To try this sample
+
+### Register your Teams Auth SSO with Azure AD
+1. Register your app with Microsoft identity platform via the Azure AD portal
+   - Your app must be registered in the Azure AD portal. See [Register an application with the Microsoft identity platform](https://docs.microsoft.com/en-us/graph/auth-register-app-v2).
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/README.md:46]: Do we need Auth SSO for this sample?

**Review by v-chetsh — CHANGES_REQUESTED** [review]
Reviewed and provided my feedback.

File: README.md [hunk 19]
```diff
@@ -206,7 +208,8 @@ The [Teams Toolkit](https://marketplace.visualstudio.com/items?itemName=TeamsDev
 [sequential#workflow#js]:samples/bot-sequential-flow-adaptive-cards/nodejs
 [app-auth#ts]:samples/app-auth/nodejs
 [auth-in-connector#nodejs]:samples/connector-todo-notification/nodejs
-[botwithsharepointfileviewer#csharp]:sample/bot-sharepoint-file-viewer/csharp
+[botwithsharepointfileviewer#csharp]:samples/bot-sharepoint-file-viewer/csharp
+[typeaheadsearch#csharp]:samples/type_ahead_search_adaptive_cards/csharp
```
Comment (v-chetsh) [inline README.md:212]: Why underscore (_), we have used hyphen (-) for all other sample names ?

File: samples/type_ahead_search_adaptive_cards/csharp/README.md [hunk 20]
```diff
@@ -0,0 +1,94 @@
+---
+page_type: sample
+description: Demonstrating the feature of type ahead search (static and dynamic) control in Adaptive Cards.
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/README.md:3]: control on Adaptive Cards

File: samples/type_ahead_search_adaptive_cards/csharp/README.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+- office-365
+languages:
+- nodejs
+extensions:
+contentType: samples
+createdDate: "11-11-2021 23:30:17"
+---
+# Type ahead search control on Adaptive Cards
+
+This sample shows the feature of type ahead search (static and dynamic) control in Adaptive Cards.
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/README.md:16]: control on Adaptive Cards

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+  "id": "{{Microsoft-App-Id}}",
+  "packageName": "com.microsoft.teams.typeaheadsearchcard",
+  "developer": {
+    "name": "Microsoft",
+    "websiteUrl": "https://www.microsoft.com",
+    "privacyUrl": "https://www.microsoft.com/privacy",
+    "termsOfUseUrl": "https://www.microsoft.com/termsofuse"
+  },
+  "name": {
+    "short": "Typeahead search Adaptive Card",
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json]: Typeahead search on Adaptive Card

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+    "full": "Adaptive card with search control"
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json]: Adaptive Card with type ahead search control

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+  },
+  "description": {
+    "short": "Typeahead search Adaptive Card",
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/AppPackage/manifest.json]: Typeahead search on Adaptive Card

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+                var jsonString = JsonConvert.SerializeObject(data);
+                JObject jsonData = JObject.Parse(jsonString);
+
+                adaptiveCardResponse = new InvokeResponse()
+                {
+                    Status = 200,
+                    Body = jsonData
+                };
+
+                    return adaptiveCardResponse;
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs]: Alignment is not correct

File: samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+        protected override async Task<InvokeResponse> OnInvokeActivityAsync(ITurnContext<IInvokeActivity> turnContext, CancellationToken cancellationToken)
+        {
+            InvokeResponse adaptiveCardResponse;
+            if (turnContext.Activity.Name == "application/search")
+            {
+                var searchData = JsonConvert.DeserializeObject<DynamicSearchCard>(turnContext.Activity.Value.ToString());
+                var packageResult = JObject.Parse(await (new HttpClient()).GetStringAsync($"https://azuresearch-usnc.nuget.org/query?q=id:{searchData.queryText}&prerelease=true"));
+                var packages = packageResult["data"].Select(item => (item["id"].ToString(), item["description"].ToString()));
+                var packageList = packages.Select(item => { var obj = new { title = item.Item1, value = item.Item2 }; return obj; }).ToList();
+                var data = new
```
Comment (v-chetsh) [inline samples/type_ahead_search_adaptive_cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs]: What is data variable here, please add meaningful name

**Review by v-chetsh — APPROVED** [review]
Changes looks fine, approving the PR.

File: samples/type-ahead-search-adaptive-cards/csharp/README.md [hunk 27]
```diff
@@ -0,0 +1,94 @@
+---
+page_type: sample
```
Comment (Wajeed-msft) [inline samples/type-ahead-search-adaptive-cards/csharp/README.md:2]: Let's follow the correct naming convention.  Please update the folder and make sure changes are reflected in ReadMe and Deployment scripts.

samples/bot-type-ahead-search-adaptive-cards

File: samples/type-ahead-search-adaptive-cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+                var packages = packageResult["data"].Select(item => (item["id"].ToString(), item["description"].ToString()));
+                var packageList = packages.Select(item => { var obj = new { title = item.Item1, value = item.Item2 }; return obj; }).ToList();
+                var searchResponseData = new
+                {
+                    type = "application/vnd.microsoft.search.searchResponse",
+                    value = new
+                    {
+                        results = packageList
+                    }
+                };
```
Comment (Wajeed-msft) [inline samples/type-ahead-search-adaptive-cards/csharp/TypeaheadSearch/Bots/ActivityBot.cs:101]: Let's check handle following scenarios.

1. No results  (204). 
2. Internal Server Errors (500)

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

File: samples/bot-type-ahead-search-adaptive-cards/csharp/README.md [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+- office-365
+languages:
+- C#
+extensions:
+contentType: samples
+createdDate: "11-11-2021 23:30:17"
+---
+# Type ahead search control on Adaptive Cards
+
+This sample shows the feature of type ahead search (static and dynamic) control on Adaptive Cards.
```
Comment (v-chetsh) [inline samples/bot-type-ahead-search-adaptive-cards/csharp/README.md:16]: We did not mentioned that how are we going to get these static or dynamic cards??

File: samples/bot-type-ahead-search-adaptive-cards/csharp/README.md [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+
+`Static search`
```
Comment (v-chetsh) [inline samples/bot-type-ahead-search-adaptive-cards/csharp/README.md:18]: Also explain a bit more on how static and dynamic search works?

File: samples/bot-type-ahead-search-adaptive-cards/csharp/README.md [hunk 31]
```diff
@@ -0,0 +1,94 @@
...
+
+![static search card](TypeaheadSearch/Images/staticSearchCard.png)
+
+`Dynamic search`
+
+![dynamic search card](TypeaheadSearch/Images/dynamicSearchCard.png)
```
Comment (v-chetsh) [inline samples/bot-type-ahead-search-adaptive-cards/csharp/README.md:24]: We did not mentioned that how are we going to get these static or dynamic cards??

File: samples/bot-type-ahead-search-adaptive-cards/csharp/README.md [hunk 32]
```diff
@@ -0,0 +1,94 @@
...
+---
+# Type ahead search control on Adaptive Cards
+
+This sample shows the feature of type ahead search (static and dynamic) control on Adaptive Cards.
+
+`Static search`
+
+![static search card](TypeaheadSearch/Images/staticSearchCard.png)
+
+`Dynamic search`
```
Comment (v-chetsh) [inline samples/bot-type-ahead-search-adaptive-cards/csharp/README.md:22]: Also explain a bit more on how static and dynamic search works?

File: samples/bot-type-ahead-search-adaptive-cards/csharp/README.md [hunk 33]
```diff
@@ -0,0 +1,94 @@
...
+
+![dynamic search card](TypeaheadSearch/Images/dynamicSearchCard.png)
+
+`Dynamic search results`
```
Comment (v-chetsh) [inline samples/bot-type-ahead-search-adaptive-cards/csharp/README.md:26]: We have "Submit" button on card so what will happen once we submit it??
Mention in doc for it.

**Review by v-chetsh — CHANGES_REQUESTED** [review]
Reviewed and provided my comments.

**Review by v-chetsh — APPROVED** [review]
Changes looks fine, approving the PR.

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**